### PR TITLE
[master] Train 212

### PR DESCRIPTION
**Pull request by falseperfection**
#1833 - Update package/cosmos api usage in integration tests
#1866 - Admin Router: AR-NEXT #4

### CODE REVIEW

**Comment by falseperfection** [comment]
@mesosphere-mergebot bump-ee

**Comment by orsenthil** [comment]
@falseperfection  - Not your fault .  For the merge-commits in Admin Router PR, I had made a comment for @vespian  here previously: https://github.com/dcos/dcos/pull/1802#issuecomment-322583779 

PR #1866  still has those. :(

Here is my suggestion for @vespian .

1. Do a rebase instead of merge when working on branches so that your changes are always on top.
2. When submitted the changes to dcos/dcos from AR-next branch rebase -i master (and squash them to a single commit). 

This will present a much better picture of dcos/dcos history.

**Comment by gpaul** [comment]
@orsenthil PRs from the ar-next branch are trains. Trains are intended to merge multiple smaller PRs that were approved in isolation.

**Comment by gpaul** [comment]
@orsenthil perhaps if AdminRouter was a separate repo smaller PRs would all be approved and applied to that repo and this repo would see only regular 'bump' PRs. 

As it stands AdminRouter (like, for example, `gen`) are part of the main repo. 

The fact that we have regular AdminRouter PRs is due to the fact that it is under active development. If `gen` was under active development we would have the same situation.

The solution is to wait for AdminRouter development to settle or if that's going to take too long we need to split it into it's own repo again.

The first option is somewhat annoying, I'll concede. However, the second is a large chunk of work that will significantly impact other deliverables for the 1.11 cycle.

@vespian do you agree with this assessment? Do you expect AdminRouter development to settle within the next few weeks?

**Comment by vespian** [comment]
> @vespian do you agree with this assessment? Do you expect AdminRouter development to settle within the next few weeks?

@gpaul I agree, and I expect that AR will settle, but only for a brief moment. Even though AR is in a really great shape now, a major API refactoring is in being slowly discussed and planned:
* https://docs.google.com/document/d/13RoNSBpS8sTaSNmwj9_rctRY7xUhZWV07Fe96elKEow/edit#heading=h.sqzdnqu3w8l5
* https://jira.mesosphere.com/browse/DCOS-5741
* https://jira.mesosphere.com/browse/DCOS_OSS-1432
* https://jira.mesosphere.com/browse/DCOS_OSS-1198
* https://jira.mesosphere.com/browse/DCOS_OSS-1441
* https://jira.mesosphere.com/browse/DCOS_OSS-1055
* https://jira.mesosphere.com/browse/DCOS_OSS-1067

On top of that, there are features that we would like to land at some point, but they have not been prioritised for any specific release:
* https://jira.mesosphere.com/browse/DCOS-11487 `Add service accounts support to AdminRouter`
* https://jira.mesosphere.com/browse/DCOS-9032 `Adminrouter option to not rewrite URLs`
* authz query caching - there is no Jira for this yet, but it was already discussed on few occacions

And finally - Admin Router is one of the central DC/OS components, it directly or indirectly connects to almost all services running in DC/OS. As long as these components keep changing, Admin Router will continue to be adjusted/refactored/updated - this work will never end.

Reply to @orsenthil 's comment will follow.

**Comment by vespian** [comment]
@orsenthil Below you can find some argument why I think we shouldn't squash commits and merge these branches as-is.

First of all I would like to stress out what @gpaul already said - the AR-NEXT PRs are in fact trains. If you check the PR descriptions, you will discover that:
* Open DC/OS AR-NEXT PR consists of 7 different PR:
  * https://github.com/mesosphere/dcos/pull/41
  * https://github.com/mesosphere/dcos/pull/42
  * https://github.com/mesosphere/dcos/pull/43
  * https://github.com/mesosphere/dcos/pull/44
  * https://github.com/mesosphere/dcos/pull/45
  * https://github.com/mesosphere/dcos/pull/46
  * https://github.com/mesosphere/dcos/pull/47
* EE DC/OS AR-NEXT PR consists of 6 PRs - some of them related to Open DC/OS PRs, some are independent:
  * https://github.com/mesosphere/dcos-enterprise/pull/1346
  * https://github.com/mesosphere/dcos-enterprise/pull/1360
  * https://github.com/mesosphere/dcos-enterprise/pull/1299
  * https://github.com/mesosphere/dcos-enterprise/pull/1319
  * https://github.com/mesosphere/dcos-enterprise/pull/1361
  * https://github.com/mesosphere/dcos-enterprise/pull/1366

These PRs address following **11 independent** JIRA issues:
- https://jira.mesosphere.com/browse/DCOS_OSS-1572 `Admin Router: Replace JWT with `DC/OS authentication token` in AR docs`
- https://jira.mesosphere.com/browse/DCOS_OSS-1487 `Admin Router: configure for Mesos V1 Streaming API`
- https://jira.mesosphere.com/browse/DCOS-14563 `Admin Router: switch /metadata to "allow all authenticated users" instead of ACL lookup`
- https://jira.mesosphere.com/browse/DCOS-9657 `Admin Router: update to openresty/nginx 1.11.x`
- https://jira.mesosphere.com/browse/DCOS-13534 `Expose DC/OS Agent version info from the 'dcos-metadata/dcos-version.json' endpoint`
- https://jira.mesosphere.com/browse/DCOS_OSS-1227 `Admin Router: Make /service endpoint logic go directly to Mesos /state-summary data for Root Marathon and Metronome`
- https://jira.mesosphere.com/browse/DCOS-16017 `Admin Router: clean up use of cert in Open AR devkit container`
- https://jira.mesosphere.com/browse/DCOS-18241 `Admin Router: Synchronize certificates used by EE and Open AR test harness`
- https://jira.mesosphere.com/browse/DCOS-17327 `Admin Router: add tests for CockroachDB endpoints`
- https://jira.mesosphere.com/browse/DCOS-16017 `Admin Router: clean up use of cert in Open AR devkit container`
- https://jira.mesosphere.com/browse/DCOS-18242 `Admin Router: Remove unnecessary fixtures`

So squashing these commits would create one big commit with 11 independent changes, I am not sure how this commit should be named, and in the case when there are bugs - how we can bisect offending change and revert it without reverting ten other changes.

I also very much liked the comment @adam-mesos one made: https://github.com/dcos/dcos/pull/1534#issuecomment-300990323 when we were discussing the very same issue a few months ago for https://github.com/dcos/dcos/pull/1534. Quoting him here:
```
Chiming on the squash convo, I don't think there's a real need to squash here, as there can be valuable information in breaking up the AR commits. Some guidelines I like to follow:

    If # commits > # files, you should consider squashing (not true here)
    If # commits / PR > 10, you should consider squashing or splitting the PR (only true for #1438 )
    No trivial commit messages like "Fixed tpyo" or "speling" (these are pretty good)
    Thankfully, each commit is prefixed with "Admin Router:" for easy scanning/filtering.
```
I find it as a very nice rule of thumb on when we should and shouldn't squash commits. In case of this PR we have:
* 9 non-merge commits  in case of Open DC/OS PR, with 11 issues in total addressed by both EE and Open PRs
* 9 non-merge commits  in case of Open DC/OS PR, with 21 files changed in total
* 1-2 commits per individual PR in case of Open DC/OS PR
* no commits like "typo" or "spelling"

I hope that I managed to convince you that squashing non-merge PRs does not make much sense in case of AR-NEXT trains. In my next comment I will try to justify why the same applies for merge commits.


The Open DC/OS AR-NEXT PR, excluding Merge commits, contains So the logical conclusion is that

**Comment by vespian** [comment]
@orsenthil In my humble opinion, the way we handle merge commits will shape the way we want to structure PR reviews and the way we handle accountability for our DC/OS repos.

In the following text, I will be referring to `a)`, `b)`, `c)`, `d)`, etc... paragraphs in order to show the logical connection between them.

a) `mesosphere-mergebot` currently removes `ShipIt` label if the branch HEAD's SHA changes. What is more, the `mergebot/review/approved/min_2` check will not go green unless the reviews were made against the current HEAD of the PR branch. Please correct me if this is not the case. My understanding is that we want the PR reviews to reflect exactly the state that is going to be merged into DC/OS and we want to prevent the malicious actor from i.e. pushing the changes into the branch after it has been peer-reviewed.

b) because of the fact that DC/OS integration process in unsuitable for fast-changing components like Admin Router, we were forced to introduce intermediate PRs against which we integrate - AR-NEXT branches. Each individual PR is peer-reviewed and then merged into the `ar-next` branch which in turn gets merged into `master`. Forbidding rebasing/keeping merge commits intact allows the reviewers to just compare the SHA sums of the `ar-next` branch and merge commits during the final review of the AR-NEXT PR, instead  of re-reviewing the whole merge branch once again, which at this stage is usually thousands of lines long.

c) approach described in `b)` also prevents a malicious actor scenario. Without having SHAsums intact, it is easy to piggyback some small change on top ot the changes from individual PRs right before the final review of AR-NEXT PRs. The final review of the AR-NEXT will never be deep enough to detect it - we can't reasonably expect reviewers to go through a ~1-3k LOC PR containing multiple independent changes and have them provide a thorough review.

d) If we strictly adhere to 1-2 commits per DC/OS PR, we will also make splitting changes into smaller PRs a wasted *and* unnecessary effort. What is the point of splitting the PR into smaller pieces if in the end, reviewers need to review the big PR with all the changes in anyway?

e) if we agree that argument `c)` is not an issue, then we should probably drop the mergebot functionality described in `a)`. Either there is an end-to-end trust, or there is none. Having strict accountability only at the end of the merge process gives no accountability at all.

f) disregarding the concerns from `b)` and forcing Admin Router maintainers to *always* submit a separate PR for even the smallest change will be consistent with `a)` while still preventing scenario `b)`. But this will severely cripple our development speed. AR PRs often overlap, and we would have to very often limit ourselves to just 1 PR per DC/OS train, then rebase, re-run tests, ask repo owners for `ShipIT`, then merge, and repeat for each and every change. I think that I do not even have to tell you how painful it is to do it now. On top of that, the quality of PRs will suffer as people will try hard to ship as much as possible in one PR.

## Summary
My suggestion is that we embrace the full accountability from paragraph `a)` and extend it to subTrains like e.g. AR-NEXT. 

By forbidding rebases during PR review we provide honest GIT history which accurately describes the history of the change. This results in less-than-optimal commit history but also a way less painful development process and great auditing of each and every change we introduce.

This approach may apply to any other component that is a part of DC/OS repo  - either now or in the future.

**Comment by gpaul** [comment]
Thanks @vespian. I disagree with only one sentiment:

> This results in less-than-optimal commit history

Why? It's a perfect commit history.

Its granular commit history is as important as any other component's commit history. The fact that its commit history happens to be interleaved with the rest of the dcos repo's commit history is a clear artefact of AdminRouter not having it's own repository.

The reason it feels awkward is that @orsenthil wants a split repository while what we have (with AR and `gen`) is a mono-repo. Mono-repo's have exactly this kind of commit history. In fact they have *only* the kind of commit history that AdminRouter is generating and have *no* 'bump PRs'.

We have two choices: split AR into a separate repository, or live with the fact that what are essentially internal AR commits end up in the main dcos commit log. Either option is fine, just priotize the ensuing efforts against other work and prioritize accordingly.

**Comment by orsenthil** [comment]
Hi @vespian  / @gpaul  - Thanks for engaging, trying to understand my motivations and presenting your situation.  I understand your state and in fact, I also had an extensive discussion with Ben Whitehead on this, who was perfectly aligned with what you both stated above.

Some premise - I had experience with large monorepos (twitter), and over there, we went with squashing changes before merging (The squash was done by a tool at remote before merge). The idea was, given a mono-repo, if components brought in small-commits and merges etc, following the history will be hard to follow. Each commit was associated with a JIRA for a bug/feature etc. 

And we need not go far, mesos (https://git-wip-us.apache.org/repos/asf?p=mesos.git;a=log) every commit is a single review-board PR and notice no merge commits there.

In `dcos/dcos`, we have accepted merge-commits for the `PR`s for the train process. Pawel explained on why we do that in a, b and c. 

Should we follow the entire (a, b, c) for AR-NEXT train too?  I was not sure. Pawel, on the points you raised, the only thing I had not thought about was *c*. I assumed  that AR-NEXT PR once reviewed could be rebased+merged on top each other, because it was entirely your ownership.

Given the points you raised,  If are you merging changes to `prozlach/ar-next` at `mesosphere/dcos` using Github, perhaps you could set Github options to like this.

image 

And use the Github's (a tool's) Rebase facility (3rd option) to land your changes on that branch, when you merge?  The same could be done by command line too, I am fine with it, but you raised your points for not doing it in c above.

I believe, all the arguments you made are still satisfied by this approach above.

**What do we gain?**

When you present the next `dcos/dcos` PR (to this mono repo), the PR will essentially not have `sub-merge` commits of AR-NEXT branch, and perhaps our merge commits are left with only the ones landing in `dcos/dcos`

My request is, not for this train, but for the next AR-NEXT dcos/dcos PR, shall we "try" the above and see if it bring's real any benefits?  

I understand that we could live with merge-commits as a fact, but given example (mesos) of how projects have avoided those, I am tending to gravitate towards the idea that - can we do it too, or  can we at-least keep these to a minimum, etc. 

So I will remove my objection here. I hope you will be willing to a try for the AR-NEXT branch PR as I shared above. (And yeah, if that is impossible, please raise it).

**Comment by gpaul** [comment]
Thanks, TIL

**Comment by lingmann** [comment]
@mesosphere-mergebot bump-ee
